File: py/tests/cross_validation/python/dump_curl.py
#!/usr/bin/env python3
"""Python-side raw-value dump matching cpp/dump_curl.cc, for cross-validating
py.vector_operations.Curl against the C++ reference (see the .cc file's
header comment for build/run instructions and py/tests/cross_validation/
README.md for the overall workflow).

Run from the repo root:  python3 -m py.tests.cross_validation.python.dump_curl
"""

from py.grid import Grid
from py.flux import Flux
from py.scalar import Scalar
from py.direction import Direction
from py.vector_operations import Curl
from .inner_products import InnerProductFlux, InnerProductScalar


def AdjointOfScalarCurl(q_, f):
    for lev in range(q_.Ngrid()):
        for i in range(1, q_.Nx()):
            for j in range(1, q_.Ny()):
                e = Scalar(q_.getGrid())
                e.assign(0.0)
                e._data[lev][i - 1, j - 1] = 1.0
                curlE = Flux(q_.getGrid())
                Curl(e, curlE)
                f._data[lev][i - 1, j - 1] = InnerProductFlux(q_, curlE) / InnerProductScalar(e, e)


def dumpScalar(name, s):
    print(name)
    for lev in range(s.Ngrid()):
        for i in range(1, s.Nx()):
            for j in range(1, s.Ny()):
                print(repr(float(s._data[lev][i - 1, j - 1])))


def dumpFlux(name, q_):
    print(name)
    for lev in range(q_.Ngrid()):
        for i in range(0, q_.Nx() + 1):
            for j in range(0, q_.Ny()):
                print(repr(q_(lev, Direction.X, i, j)))
        for i in range(0, q_.Nx()):
            for j in range(0, q_.Ny() + 1):
                print(repr(q_(lev, Direction.Y, i, j)))


def main():
    nx, ny, ngrid = 8, 8, 2
    length = 8.0
    grid = Grid(nx, ny, ngrid, length, -length / 2, -length * ny / nx / 2)
    u = Scalar(grid)
    v = Scalar(grid)
    q = Flux(grid)
    u.assign(0.0)
    v.assign(0.0)
    q.assign(0.0)

    for lev in range(ngrid):
        for i in range(1, nx):
            for j in range(1, ny):
                u._data[lev][i - 1, j - 1] = grid.getXEdge(lev, i)
                v._data[lev][i - 1, j - 1] = grid.getYEdge(lev, j)

    for lev in range(ngrid):
        dx = grid.Dx(lev)
        for i in range(0, nx + 1):
            for j in range(0, ny):
                q.set(lev, Direction.X, i, j, value=-2 * q.y(lev, Direction.X, j) * q.y(lev, Direction.X, j) * dx)
        for i in range(0, nx):
            for j in range(0, ny + 1):
                q.set(lev, Direction.Y, i, j, value=q.x(lev, Direction.Y, i) * q.x(lev, Direction.Y, i) * dx)

    curlU = Curl(u)
    curlV = Curl(v)
    omega = Curl(q)

    omega2 = Scalar(grid)
    omega2.assign(0.0)
    AdjointOfScalarCurl(q, omega2)
    err = Scalar(grid)
    err.assign(omega)
    err._data[...] = omega._data - omega2._data

    dumpScalar("DUMP_u", u)
    dumpFlux("DUMP_curlU", curlU)
    dumpScalar("DUMP_v", v)
    dumpFlux("DUMP_curlV", curlV)
    dumpFlux("DUMP_q", q)
    dumpScalar("DUMP_omega", omega)
    dumpScalar("DUMP_omega2", omega2)
    dumpScalar("DUMP_err", err)


if __name__ == "__main__":
    main()

File: py_static/tests/cross_validation/cpp/dump_curl.cc
// dump_curl.cc
//
// Raw-value dump driver for cross-validating py/vector_operations.py's Curl
// against the C++ reference implementation. Reproduces the exact scenario
// from test/CheckCurl.cc (same grid, same analytic fields), but prints one
// full-precision value per line under a "DUMP_<name>" marker instead of
// Scalar::print()/Flux::print()'s fixed-width grid layout, so the output is
// trivially diffable against python/dump_curl.py's output via compare_dumps.py.
//
// Build (from repo root):
//   g++ -Wall -O2 -I src -c py/tests/cross_validation/cpp/dump_curl.cc -o /tmp/dump_curl.o
//   g++ /tmp/dump_curl.o -L build -libpm -lfftw3 -lm -o /tmp/dump_curl
// Run:
//   /tmp/dump_curl > /tmp/dump_curl.out

#include <iostream>
#include <iomanip>
#include "ibpm.h"

using namespace std;
using namespace ibpm;

void AdjointOfScalarCurl( const Flux& q, Scalar& f ) {
    for (int lev=0; lev<q.Ngrid(); ++lev) {
        for (int i=1; i<q.Nx(); ++i) {
            for (int j=1; j<q.Ny(); ++j) {
                Scalar e(q.getGrid());
                e = 0;
                e(lev,i,j) = 1.;
                Flux curlE(q.getGrid()) ;
                Curl( e, curlE );
                f(lev,i,j) = InnerProduct( q, curlE ) / InnerProduct( e, e );
            }
        }
    }
}

void dumpScalar(const char* name, const Scalar& s) {
    cout << name << endl;
    for (int lev=0; lev<s.Ngrid(); ++lev)
        for (int i=1; i<s.Nx(); ++i)
            for (int j=1; j<s.Ny(); ++j)
                cout << setprecision(17) << s(lev,i,j) << endl;
}
void dumpFlux(const char* name, const Flux& q) {
    cout << name << endl;
    for (int lev=0; lev<q.Ngrid(); ++lev) {
        for (int i=0; i<=q.Nx(); ++i)
            for (int j=0; j<q.Ny(); ++j)
                cout << setprecision(17) << q(lev,X,i,j) << endl;
        for (int i=0; i<q.Nx(); ++i)
            for (int j=0; j<=q.Ny(); ++j)
                cout << setprecision(17) << q(lev,Y,i,j) << endl;
    }
}

int main() {
    int nx=8, ny=8, ngrid=2;
    double length=8;
    Grid grid( nx, ny, ngrid, length, -length/2, -length*ny/nx/2 );
    Scalar u(grid);
    Scalar v(grid);
    Flux q(grid);

    for (int lev=0; lev<ngrid; ++lev) {
        for (int i=1; i<nx; ++i) {
            for (int j=1; j<ny; ++j) {
                u(lev,i,j) = grid.getXEdge(lev,i);
                v(lev,i,j) = grid.getYEdge(lev,j);
            }
        }
    }

    for (int lev=0; lev<ngrid; ++lev) {
        double dx = grid.Dx(lev);
        for (int i=0; i<=nx; ++i)
            for (int j=0; j<ny; ++j)
                q(lev,X,i,j) = -2*q.y(lev,X,j)*q.y(lev,X,j) * dx;
        for (int i=0; i<nx; ++i)
            for (int j=0; j<=ny; ++j)
                q(lev,Y,i,j) = q.x(lev,Y,i)*q.x(lev,Y,i) * dx;
    }

    Flux curlU(grid);
    Flux curlV(grid);
    Curl( u, curlU );
    Curl( v, curlV );

    Scalar omega(grid);
    Curl( q, omega );
    Scalar omega2(grid);
    AdjointOfScalarCurl( q, omega2 );
    Scalar err = omega - omega2;

    dumpScalar("DUMP_u", u);
    dumpFlux("DUMP_curlU", curlU);
    dumpScalar("DUMP_v", v);
    dumpFlux("DUMP_curlV", curlV);
    dumpFlux("DUMP_q", q);
    dumpScalar("DUMP_omega", omega);
    dumpScalar("DUMP_omega2", omega2);
    dumpScalar("DUMP_err", err);
    return 0;
}
